### PR TITLE
Pt 156095826 contract balance in account tree

### PR DESCRIPTION
**Pull request by happi**
Move contract account balances to account state tree.
Also fix path to mnesia patches during euint tests.

### CODE REVIEW

File: apps/aecontract/src/aect_contracts.erl [hunk 1]
```diff
@@ -89,7 +77,7 @@
 -define(PUB_SIZE, 32).
 -define(HASH_SIZE, 32).
 -define(CONTRACT_TYPE, contract).
--define(CONTRACT_VSN, 1).
+-define(CONTRACT_VSN, 2).
```
Comment (lucafavatella) [inline apps/aecontract/src/aect_contracts.erl]: I expect you can keep 1 (cc @gorillainduction ) as anyway we [plan to bump consensus](https://www.pivotaltracker.com/story/show/157506836)

File: apps/aecontract/src/aect_contracts.erl [hunk 2]
```diff
@@ -142,11 +125,7 @@ serialize(#contract{} = C) ->
       ?CONTRACT_TYPE,
       ?CONTRACT_VSN,
       serialization_template(?CONTRACT_VSN),
-      [ {pubkey, pubkey(C)}
```
Comment (lucafavatella) [inline apps/aecontract/src/aect_contracts.erl]: Could you please make the environments start from a fresh DB, by [bumping the DB folder](https://github.com/aeternity/epoch/blob/10b0a4e9a03d178b0d6fed470d7b246b2ae876c0/deployment/ansible/deploy.yml#L69)?

Comment (lucafavatella) [inline apps/aecontract/src/aect_contracts.erl]: Could you please propose contextual update of [serialization doc](https://github.com/aeternity/protocol/blob/414051131a29b808bd829525f773e8fc3ae75d2d/serializations.md#contract)?

File: apps/aecontract/src/aect_create_tx.erl [hunk 3]
```diff
@@ -128,13 +128,26 @@ nonce(#contract_create_tx{nonce = Nonce}) ->
...
+                          gas_price  = GasPrice,
                           fee = Fee}, _Context, Trees, Height, _ConsensusVersion) ->
+    TotalAmount = Fee + Amount + Gas * GasPrice,
     Checks =
-        [fun() -> aetx_utils:check_account(OwnerPubKey, Trees, Height, Nonce, Fee) end],
-
+        [fun() -> aetx_utils:check_account(OwnerPubKey, Trees, Height, Nonce, TotalAmount) end,
+         %% TODO: Check minum gas price.
+         fun () ->
+                 case Fee >= aec_governance:minimum_tx_fee() of
```
Comment (lucafavatella) [inline apps/aecontract/src/aect_create_tx.erl]: Is this not already checked by [tx generic code](https://github.com/aeternity/epoch/blob/10b0a4e9a03d178b0d6fed470d7b246b2ae876c0/apps/aecore/src/aec_trees.erl#L191)?

File: apps/aecontract/src/aect_create_tx.erl [hunk 4]
```diff
@@ -128,13 +128,26 @@ nonce(#contract_create_tx{nonce = Nonce}) ->
...
+                          nonce = Nonce,
+                          amount     = Amount,
+                          gas        = Gas,
+                          gas_price  = GasPrice,
                           fee = Fee}, _Context, Trees, Height, _ConsensusVersion) ->
+    TotalAmount = Fee + Amount + Gas * GasPrice,
     Checks =
-        [fun() -> aetx_utils:check_account(OwnerPubKey, Trees, Height, Nonce, Fee) end],
-
+        [fun() -> aetx_utils:check_account(OwnerPubKey, Trees, Height, Nonce, TotalAmount) end,
```
Comment (lucafavatella) [inline apps/aecontract/src/aect_create_tx.erl]: This checks that originator of tx is in accounts state tree with enough balance.
I would expect (sanity) check that computed contract pubkey is absent from account state tree.

File: apps/aecontract/src/aect_contracts.erl (hunk 2 above)
Comment (happi) [reply apps/aecontract/src/aect_contracts.erl]: https://github.com/aeternity/protocol/pull/82

File: apps/aecontract/src/aect_create_tx.erl (hunk 4 above)
Comment (happi) [reply apps/aecontract/src/aect_create_tx.erl]: There is nothing stopping you from transferring funds to a contract address before it exists (you can calculate the address beforehand) thus creating the address and adding it to the state tree. Hence I do not think we should check that the account doesn't exist.

File: apps/aecontract/src/aect_create_tx.erl [hunk 5]
```diff
@@ -190,24 +204,39 @@ process(#contract_create_tx{owner = OwnerPubKey,
 			CallsTree0 = aec_trees:calls(Trees3),
 			CallsTree1 = aect_call_state_tree:insert_call(CallRes, CallsTree0),
 			Trees4     = aec_trees:set_calls(Trees3, CallsTree1),
+
+                        %% Spend Gas
+                        GasCost = aect_call:gas_used(CallRes) * GasPrice,
+                        Trees5 =
+                            spend(OwnerPubKey, ContractPubKey, 0, GasCost, Nonce+1,
```
[the remark this answers is not in the document]
Comment (happi) [reply apps/aecontract/src/aect_create_tx.erl]: I'll take a look...

File: apps/aecontract/src/aect_contracts.erl (hunk 2 above)
Comment (lucafavatella) [reply apps/aecontract/src/aect_contracts.erl]: Thanks

File: apps/aecontract/src/aect_create_tx.erl (hunk 3 above)
Comment (gorillainduction) [reply apps/aecontract/src/aect_create_tx.erl]: It is.

**Comment by lucafavatella** [comment]
@happi 

> I could not find another way to get rebar to set the path correctly.

I locally had patch with `env ERL_LIBS="$(CURDIR)/otp_patches:$(ERL_LIBS)" ./rebar3 do eunit $(EUNIT_TEST_FLAGS)` ...

> Unfortunately the path to the patches seems to differ in the CI build. .... The euint.vm.args solved it locally, but not for CI :(

... though did not check this on CI.

**Comment by lucafavatella** [comment]
I took the liberty of marking as WIP as I understand @happi may propose further changes.

**Comment by lucafavatella** [comment]
@happi Python UAT tests failing on commit 41c3fe94 - see [helpful auxiliary CI job](https://circleci.com/gh/lucafavatella/epoch/3593).

File: apps/aecontract/src/aect_contracts.erl (hunk 2 above)
Comment (lucafavatella) [reply apps/aecontract/src/aect_contracts.erl]: Somehow lost in rebase?

**Comment by lucafavatella** [comment]
In [CI test job ](https://circleci.com/gh/aeternity/epoch/12755) I see socket_closed_remotely - I am restarting job.

**Comment by lucafavatella** [comment]
> I am preparing separate PR addressing Python UAT tests.

@happi PR addressing Python UAT tests https://github.com/aeternity/epoch/pull/1148

Up to you whether you prefer to merge that 1148 to this PR 1131 and then merge to master in one go, or rather merge this 1131 to master first and then merge 1148 to master later.

**Comment by lucafavatella** [comment]
> In CI test job I see socket_closed_remotely - I am restarting job.

I [restarted CI test job](https://circleci.com/gh/aeternity/epoch/12787) I saw same error again - I restarted again.

File: apps/aecore/src/aec_vm_chain.erl [hunk 6]
```diff
@@ -160,54 +156,21 @@ do_set_store(Store, PubKey, Trees) ->
...
-            {account, _}  -> do_spend_to_account(Recipient, ContractKey, Amount, Trees, Height);
-            none          -> do_spend_to_account(Recipient, ContractKey, Amount, Trees, Height)
-        end
-    catch throw:bad_recip  -> {error, {bad_recipient_account, Recipient}};
-          throw:bad_height -> {error, {account_height_too_big, Recipient}};
-          throw:no_funds   -> {error, insufficient_funds};
-          _:_              -> {error, unspecified_error}    %% TODO
+    AccountTree = aec_trees:accounts(Trees),
+    {value, Account} = aec_accounts_trees:lookup(ContractKey, AccountTree),
+    Nonce = aec_accounts:nonce(Account) + 1,
```
[the remark this answers is not in the document]
Comment (gorillainduction) [reply apps/aecore/src/aec_vm_chain.erl]: It is a bit strange that the nonce of the contract is bumped, without any transaction recorded on the chain, but that might be the way contracts should act.

**Pull request merged**